### DISCUSSION TITLE
How can I use an API call to describe an object with a list of unknown size as one of its parameters?

### DISCUSSION THREAD

**Discussion by Disciple153** [aws/aws-cdk#23852] [Q&A]
I'm working on adding QuickSight support, and as part of that I'm trying to make the ability to import existing QuickSight resources, but a snag that I've come across is that in order to create a Template from a Analysis, I need to input a list of  DataSet Arns from the Analysis into a differently formatted list for the Template:

QuickSight::DescribeAnalysis:
```json
{
  "DataSets": [
    {
      "DataSetArn": "arn"
    }
  ]
{
```
CfnTemplate.sourceEntity:
```json
{
  "sourceAnalysis": {
    "dataSetReferences": [
      {
        "dataSetArn": "arn",
        "dataSetPlaceholder": "arn"
      }
    ]
  }
}
```

My problem is that if I use a CustomResource to get the AnalysisDescription, there is no way for me to tell how many DataSets the Analysis will have. I could use the aws-sdk, but I doubt that will get through a PR.
[This article](https://garbe.io/blog/2019/09/20/hey-cdk-how-to-use-existing-resources/) implies that there is a way to make API calls at runtime, but I can't figure out how to do that.

Here is [my fork](https://github.com/Disciple153/aws-cdk/tree/dev/packages/%40aws-cdk/aws-quicksight) if it's of any help.

**Comment by OwaisAzad** [comment]
One solution to your problem could be to write a custom resource in your CloudFormation template that loops through the DataSets in the Analysis description and creates the appropriate number of dataSetReferences in the sourceEntity for the Template.

**Reply by Disciple153** [reply]
That could work, but it would be a messy solution and would leave many custom resources in the users' CloudFormation deployments.

After talking to the CDK team, the correct solution is to create ContextProviders the same way that  Vpc.fromLookup() does. I have not gotten this working yet, but when I do, I will update this post with my answer.
